Commit for Moka-001 , Moka-002 , Moka-003, Moka-004

nazmy opened the pull request:
Hi Grady,

I have completed Moka-001 to Moka-004 development. I'm very grateful to have the opportunity to do this Interview test, it really encourages me to learn and master RoR. Frankly speaking, I had never seen Ruby code or even Installed RoR before this test. I had zero knowledge of Ruby on Monday afternoon, but I'm very happy that I could pick up Ruby and develop fully working portal with automated testing in short time.

I really learned a lot when doing Moka-001 to Moka-003 from RVM, Ruby, Gems, Rails Framework, Rake, Controller, Models, Views, Callback, Test and a lot of other things. 

It took me a pause prior doing Moka-004. In Moka-0004 development, I created role model, roles and implemented cancancan authorisation. By using cancancan, we can easily manage the privileges of different role. Besides cancancan, I was thinking to implement devise authentication to replace the existing authentication. However, I don't think it's required for now and I don't have enough time to try to explore and implement it.

There are other enhancement that can be done as below:
1) Modifies relation user-business to many-to-many, since one business can be managed by multiple user and one user can manage multiple business. The user may have different role for each business. Alternatively, create branches with many-to-one relation to business.
2)  Added 'Admin' and 'Seller' User registration feature for authenticated 'Admin' User. Business Admin User, shall be responsible of creating Seller and other Admin Users.
3) Remove 'Business' from sign-up page since sign-up shall only be for regular user. 
4) Use devise authentication
...and much more enhancements..

Hope you are satisfied with the result. Look forward to meeting you tomorrow 😄 

Cheers,
Nazmy

nazmy committed:
- Refine Item List Caption
- Implemented Add Items :
	- Modified Add button
	- Added new form
	- Create item definition
	- Implemented new and create definition

nazmy committed:
Implemented & Added Delete Method

nazmy committed:
- Added Edit Form
- Implemented Update Method

nazmy committed:
Merge branch 'MOKA-001'

* MOKA-001:
  - Added Edit Form - Implemented Update Method
  Implemented & Added Delete Method
  - Refine Item List Caption - Implemented Add Items : 	- Modified Add button 	- Added new form 	- Create item definition 	- Implemented new and create definition

nazmy committed:
- Added Business Form in New User Form
- Implemented User Create

nazmy committed:
Merge branch 'MOKA-002'

* MOKA-002:
  - Added Business Form in New User Form - Implemented User Create

nazmy committed:
- Added User Model Email Format and Uniqueness Validation
- Added length, presense validation in item and user model
- Refine User and Items Controller to handle model validation

nazmy committed:
Merge branch 'MOKA-004'

* MOKA-004:
  - Added User Model Email Format and Uniqueness Validation - Added length, presense validation in item and user model - Refine User and Items Controller to handle model validation

nazmy committed:
- Rectify business routes to 'businesses'
- Enhance header navbar to handle new vs edit business
- Added Business Controller, New Form and Edit Form
- Implemented Create Business, Edit Business
- Enhance Business Model Validation

nazmy committed:
- Added Edit User Profile
- Implemented Update User Profile

nazmy committed:
Merge branch 'MOKA-002'

* MOKA-002:
  - Added Edit User Profile - Implemented Update User Profile
  - Rectify business routes to 'businesses' - Enhance header navbar to handle new vs edit business - Added Business Controller, New Form and Edit Form - Implemented Create Business, Edit Business - Enhance Business Model Validation

nazmy committed:
- Added User Model Test
- Enhance User Model to save downcase on before_save callback

nazmy committed:
Added Business and Item Model Test

nazmy committed:
Enhance user password validation. Added password length validation and test

nazmy committed:
Added User Controller Test

nazmy committed:
- Added Test Session Helper
- Added Business Controller Test

nazmy committed:
Added Item Controller Test

nazmy committed:
Merge branch 'MOKA-003'

* MOKA-003:
  Added Item Controller Test
  - Added Test Session Helper - Added Business Controller Test
  Added User Controller Test
  Enhance user password validation. Added password length validation and test
  Added Business and Item Model Test
  - Added User Model Test - Enhance User Model to save downcase on before_save callback

nazmy committed:
Add Role and User Role Relation

nazmy committed:
- Added and Implement Cancancan
- Define Cancancan abilities
- Added Authorisation in both controller and views based on cancancan abilities

nazmy committed:
Merge branch 'MOKA-004'

* MOKA-004:
  - Added and Implement Cancancan - Define Cancancan abilities - Added Authorisation in both controller and views based on cancancan abilities
  Add Role and User Role Relation

glaksmono commented:
Thanks for the PR @nazmy ! :)

Looking forward in meeting you tomorrow. Haven't looked at this in more details, but thoughts about these?
```
➜  moka-rails-interview git:(master) rake test:all
DEPRECATION WARNING: rake test:all is deprecated and will be removed in Rails 5. Use rake test to run all tests in test directory. (called from mon_synchronize at /Users/glaksmono/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:211)
Run options: --seed 12804

# Running:

............EEEEEEEEEEE.................

Finished in 0.631523s, 63.3390 runs/s, 68.0894 assertions/s.

  1) Error:
BusinessesControllerTest#test_should_fail_create_business_if_address_is_empty:
ActiveModel::ForbiddenAttributesError: ActiveModel::ForbiddenAttributesError
    test/controllers/businesses_controller_test.rb:46:in `block in <class:BusinessesControllerTest>'


  2) Error:
BusinessesControllerTest#test_should_fail_create_business_if_city_is_empty:
ActiveModel::ForbiddenAttributesError: ActiveModel::ForbiddenAttributesError
    test/controllers/businesses_controller_test.rb:54:in `block in <class:BusinessesControllerTest>'


  3) Error:
BusinessesControllerTest#test_should_redirect_to_new:
NoMethodError: undefined method `name' for nil:NilClass
    app/models/user.rb:23:in `admin?'
    app/models/ability.rb:8:in `initialize'
    test/controllers/businesses_controller_test.rb:30:in `block in <class:BusinessesControllerTest>'


  4) Error:
BusinessesControllerTest#test_should_fail_create_business_if_name_is_empty:
ActiveModel::ForbiddenAttributesError: ActiveModel::ForbiddenAttributesError
    test/controllers/businesses_controller_test.rb:38:in `block in <class:BusinessesControllerTest>'


  5) Error:
ItemsControllerTest#test_should_update_item:
NoMethodError: undefined method `name' for nil:NilClass
    app/models/user.rb:23:in `admin?'
    app/models/ability.rb:8:in `initialize'
    test/controllers/items_controller_test.rb:89:in `block in <class:ItemsControllerTest>'


  6) Error:
ItemsControllerTest#test_should_create_item:
ActiveModel::ForbiddenAttributesError: ActiveModel::ForbiddenAttributesError
    test/controllers/items_controller_test.rb:68:in `block (2 levels) in <class:ItemsControllerTest>'
    test/controllers/items_controller_test.rb:67:in `block in <class:ItemsControllerTest>'


  7) Error:
ItemsControllerTest#test_should_get_index:
NoMethodError: undefined method `name' for nil:NilClass
    app/models/user.rb:23:in `admin?'
    app/models/ability.rb:8:in `initialize'
    test/controllers/items_controller_test.rb:41:in `block in <class:ItemsControllerTest>'


  8) Error:
ItemsControllerTest#test_should_destroy_item:
NoMethodError: undefined method `name' for nil:NilClass
    app/models/user.rb:23:in `admin?'
    app/models/ability.rb:8:in `initialize'
    test/controllers/items_controller_test.rb:102:in `block (2 levels) in <class:ItemsControllerTest>'
    test/controllers/items_controller_test.rb:101:in `block in <class:ItemsControllerTest>'


  9) Error:
ItemsControllerTest#test_should_redirect_to_new:
NoMethodError: undefined method `name' for nil:NilClass
    app/models/user.rb:23:in `admin?'
    app/models/ability.rb:8:in `initialize'
    test/controllers/items_controller_test.rb:58:in `block in <class:ItemsControllerTest>'


 10) Error:
ItemsControllerTest#test_should_redirect_to_create_business_if_business_doesnt_exists:
NoMethodError: undefined method `name' for nil:NilClass
    app/models/user.rb:23:in `admin?'
    app/models/ability.rb:8:in `initialize'
    test/controllers/items_controller_test.rb:48:in `block in <class:ItemsControllerTest>'


 11) Error:
ItemsControllerTest#test_should_get_edit:
NoMethodError: undefined method `name' for nil:NilClass
    app/models/user.rb:23:in `admin?'
    app/models/ability.rb:8:in `initialize'
    test/controllers/items_controller_test.rb:80:in `block in <class:ItemsControllerTest>'

40 runs, 43 assertions, 0 failures, 11 errors, 0 skips
➜  moka-rails-interview git:(master)
```

nazmy committed:
- Rectify load_and_authorise resource cancan
- Added role in test upon after adding role

nazmy commented:
Hi Grady,

Apologise, I forgot to update the controller testing to include Role after implementing cancancan. I was too happy last night after manage to implement cancancan. Afterwards, I was studying devise as interesting with the devise authentication.

I have rectified the controller testing and also load_and_authorize_resource. Please have a look at this PR.

Thanks,
Nazmy

nazmy committed:
- Implemented Devise to enhance authentication
- Modify routes to remove the unecessary custom session
- Replace custom authenticate_user with devise
- Remove unecessary session helper
- Modify DB for devise

nazmy committed:
Remove unecessary controller and view. Session and user controller and view are provided by Devise

nazmy committed:
Enhance Cancancan Ability and Authorisation in view and controller

nazmy committed:
- Enhance testing to test authorisation define in cancancan abilities
- Use fixtures to prepare multiple testing data

nazmy committed:
Merge branch 'MOKA-004'

* MOKA-004:
  - Enhance testing to test authorisation define in cancancan abilities - Use fixtures to prepare multiple testing data
  Enhance Cancancan Ability and Authorisation in view and controller
  Remove unecessary controller and view. Session and user controller and view are provided by Devise
  - Implemented Devise to enhance authentication - Modify routes to remove the unecessary custom session - Replace custom authenticate_user with devise - Remove unecessary session helper - Modify DB for devise
